SwiftGit2/ErrorCodes.swift

```diff
@@ -0,0 +1,319 @@
+		return (type: type, errorMessage: errorMessage)
+	}
+}
+
+/// Source Libit2 method where an error was reported
+public enum Libgit2Method {
+	case git_repository_init
+	case git_signature_new
+	case git_repository_open
+	case git_clone
+	case git_object_lookup
+	case git_remote_list
+	case git_remote_lookup
+	case git_remote_fetch
+	case git_reference_list
+	case git_reference_lookup
+	case git_repository_head
+	case git_repository_set_head
+	case git_checkout_head
+	case git_repository_index
+	case git_index_add_all
+	case git_index_write
+	case git_tree_lookup
+	case git_commit_lookup
+	case git_commit_create
+	case git_index_write_tree
+	case git_reference_name_to_id
+	case git_diff_merge
+	case git_diff_tree_to_tree
+	case git_status_init_options
+	case git_status_list_new
+	case git_repository_open_ext
+	case git_revwalk_next
+}
+
+extension git_error_code: CustomStringConvertible {
+	public var description: String {
+		switch self {
+		case GIT_OK:
+			return "No error"
+		case GIT_ERROR:
+			return "Generic error"
+		case GIT_ENOTFOUND:
+			return "Requested object could not be found"
+		case GIT_EEXISTS:
+			return "Object exists preventing operation"
+		case GIT_EAMBIGUOUS:
+			return "More than one object matches"
+		case GIT_EBUFS:
+			return "Output buffer too short to hold data"
+		case GIT_EUSER:
+			return "A special error that is never generated by libgit2"
+		case GIT_EBAREREPO:
+			return "Operation not allowed on bare repository"
+		case GIT_EUNBORNBRANCH:
+			return "HEAD refers to branch with no commits"
+		case GIT_EUNMERGED:
+			return "Merge in progress prevented operation"
+		case GIT_ENONFASTFORWARD:
+			return "Reference was not fast-forwardable"
+		case GIT_EINVALIDSPEC:
+			return "Name/ref spec was not in a valid format"
+		case GIT_ECONFLICT:
+			return "Checkout conflicts prevented operation"
+		case GIT_ELOCKED:
+			return "Lock file prevented operation"
+		case GIT_EMODIFIED:
+			return "Reference value does not match expected"
+		case GIT_EAUTH:
+			return "Authentication error"
+		case GIT_ECERTIFICATE:
+			return "Server certificate is invalid"
+		case GIT_EAPPLIED:
+			return "Patch/merge has already been applied"
+		case GIT_EPEEL:
+			return "The requested peel operation is not possible"
+		case GIT_EEOF:
+			return "Unexpected EOF"
+		case GIT_EINVALID:
+			return "Invalid operation or input"
+		case GIT_EUNCOMMITTED:
+			return "Uncommitted changes in index prevented operation"
+		case GIT_EDIRECTORY:
+			return "The operation is not valid for a directory"
+		case GIT_EMERGECONFLICT:
+			return "A merge conflict exists and cannot continue"
+		case GIT_PASSTHROUGH:
+			return "Internal only"
+		case GIT_ITEROVER:
+			return "Signals end of iteration with iterator"
+		case GIT_RETRY:
+			return "Internal only"
+		case GIT_EMISMATCH:
+			return "Hashsum mismatch in object"
+		default:
+			return "Unknown Error Code: \(self)"
+		}
+	}
+}
+
+public enum Libgit2ErrorType {
```

Comment:
Why not using the error protocol directly?
```suggestion
public enum Libgit2ErrorType: Error {
```

Reply:
At the cost of making `ErrorCodes.swift` more verbose, we can simplify calling code to the library, providing the promise that any SwiftGit2 function will only produce `SwiftGit2Error` errors (unifying errors originating from libgit2 and internally from SwiftGit2 itself).